### PR TITLE
launcher, converter: Extract Memory configuration logic

### PR DESCRIPTION
**Pull request by harshitgupta1337**
### What this PR does
This PR Extract the logic responsible for configuring guest memory to a dedicated `MemoryConfigurator`.

### References

- Partially addresses #16117

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/libvmi"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/converter/compute"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/converter/vcpu"
+)
+
+var _ = Describe("Memory Domain Configurator", func() {
+	DescribeTable("should calculate memory in bytes", func(quantity string, bytes int) {
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go:35]: The `vcpu.QuantityToByte` function is an implementation detail that is tested indirectly by the MemoryConfigurator.

Instead of testing it directly, I recommend ensuring the MemoryConfigurator tests cover the relevant scenarios (including edge cases)

**Review by orelmisan — CHANGES_REQUESTED** [review]
Thank you for the PR @harshitgupta1337! 
It is great to see more people joining this effort!

Apologies for the delay in reviewing, I needed to take a deep dive into the current code to fully understand the context.

I noticed that the existing tests in `converter_test.go` are difficult to follow, as they contain a lot of noise and unrelated details. I would like to improve that baseline before we move forward. We could handle this in one of two ways:
1. Refactor the tests as a prerequisite PR (merged before this one).
2. Rewrite the tests as part of this current PR.

Please let me know which approach you would prefer.

**Comment by harshitgupta1337** [comment]
> I noticed that the existing tests in `converter_test.go` are difficult to follow, as they contain a lot of noise and unrelated details. I would like to improve that baseline before we move forward. We could handle this in one of two ways:
> 
> 1. Refactor the tests as a prerequisite PR (merged before this one).
> 2. Rewrite the tests as part of this current PR.

Thanks for throwing light on this pain-point @orelmisan - this is something I also noticed. I would prefer Option 1, so that the test refactoring is done in an independent PR. I am looking forward to hear what you have in mind for making the test code better. I would like to propose that we move tests for different aspects of the VMI (e.g., CPU, memory) into their individual subdirectories. Common utilities will live in a separate subdirectory.
WDYT ?

**Comment by orelmisan** [comment]
> > I noticed that the existing tests in `converter_test.go` are difficult to follow, as they contain a lot of noise and unrelated details. I would like to improve that baseline before we move forward. We could handle this in one of two ways:
> > 
> > 1. Refactor the tests as a prerequisite PR (merged before this one).
> > 2. Rewrite the tests as part of this current PR.
> 
> Thanks for throwing light on this pain-point @orelmisan - this is something I also noticed. I would prefer Option 1, so that the test refactoring is done in an independent PR. I am looking forward to hear what you have in mind for making the test code better. I would like to propose that we move tests for different aspects of the VMI (e.g., CPU, memory) into their individual subdirectories. Common utilities will live in a separate subdirectory. WDYT ?

@harshitgupta1337, IMO, the existing tests in `converter_test.go` or the new MemoryDomainConfigurator tests should explicitly state the input VMI and assert equality to an expected `api.domain` object. 
The production logic here depends on:
- vmi.Spec.Domain.Memory
- vmi.Spec.Domain.Memory.MaxGuest
- vmi.Spec.Domain.Memory.Guest
- vmi.Spec.Domain.Resources.Limits[k8sv1.ResourceMemory]
- vmi.Spec.Domain.Resources.Limits[k8sv1.ResourceMemory]

so the tests should reflect the different possible scenarios of their combinations.

This logic also seem to have a potential flaw - `vmi.Spec.Domain.Memory.Guest` is not tested for nil value before it is de-referenced.

> I would like to propose that we move tests for different aspects of the VMI (e.g., CPU, memory) into their individual subdirectories.

This is one of the goals of https://github.com/kubevirt/kubevirt/issues/16117 - to have the individual aspects thoroughly tested in isolation.
I am currently in favor of keeping the converter_test.go file (at least the parts of it that do not test implementation details) as an integration test, but it too should be refactored eventually.

**Comment by harshitgupta1337** [comment]
@orelmisan Can you please take a look at this PR again? I have addressed your comments.

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go (hunk 1 above)
Comment (dasionov) [reply pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: IMO, it would be cleaner if all of these tests used quantity sizing. The additional bytes input doesn’t seem to add much value here, so we could probably simplify by removing it.

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+			configurator := compute.MemoryConfigurator{}
+			Expect(configurator.Configure(vmi, &domain)).To(Succeed())
+
+			xmlBytes, err := xml.MarshalIndent(domain.Spec, "", "  ")
+			Expect(err).ToNot(HaveOccurred())
+
+			xml := string(xmlBytes)
+
+			Expect(strings.Contains(xml, `<memory unit="b">2222222</memory>`)).To(BeTrue(), xml)
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: for consistency with other configurators tests:
```suggestion
			expectedDomain := api.Domain{Spec: api.DomainSpec{
				Memory: api.Memory{
					Value: uint64(2222222),
				},
			}}
			Expect(domain).To(Equal(expectedDomain))
```

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+				libvmi.WithGuestMemory("123Mi"),
+			)
+			v1.SetObjectDefaults_VirtualMachineInstance(vmi)
+			var domain api.Domain
+
+			configurator := compute.MemoryConfigurator{}
+			Expect(configurator.Configure(vmi, &domain)).To(Succeed())
+
+			Expect(domain.Spec.Memory.Value).To(Equal(uint64(128974848)))
+			Expect(domain.Spec.Memory.Unit).To(Equal("b"))
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]:
```suggestion

			expectedDomain := api.Domain{
				Spec: api.DomainSpec{
					Memory: api.Memory{
						Value: uint64(128974848),
						Unit:  "b",
					},
				},
			}
			Expect(domain).To(Equal(expectedDomain))
```

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+					},
+				}
+
+				domain = &api.Domain{
+					Spec: api.DomainSpec{
+						VCPU: &api.VCPU{
+							CPUs: 2,
+						},
+					},
+				}
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: Could you help clarify why the Memory configurator needs to be aware of or include vCPU information in the domain at this point? I’m trying to understand the coupling here and whether it’s intentional or incidental.

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+				v1.SetObjectDefaults_VirtualMachineInstance(vmi)
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: same here

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+		var domain api.Domain
+		configurator := compute.MemoryConfigurator{}
+		Expect(configurator.Configure(vmi, &domain)).To(Succeed())
+
+		xmlBytes, err := xml.MarshalIndent(domain.Spec, "", "  ")
+		Expect(err).ToNot(HaveOccurred())
+
+		xml := string(xmlBytes)
+		Expect(strings.Contains(xml, fmt.Sprintf("<memory unit=\"b\">%d</memory>", bytes))).To(BeTrue(), xml)
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: You can derive the bytes using: `bytes, err := resource.ParseQuantity(quantity)` this will remove the need in the second paramter.
also please dont test the xml itself, but rather the struct domain is enough
```suggestion
expectedDoamin := api.Domain{
  Spec: api.DomainSpec{
    Memory: api.Memory{
      Value: bytes.Value(),
      Unit: "b"
    },
  },
}
Expect(domain).To(Equal(expectedDomain))
```

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	},
+		Entry("specifying memory 64M", "64M", 64*1000*1000),
+		Entry("specifying memory 64Mi", "64Mi", 64*1024*1024),
+		Entry("specifying memory 3G", "3G", 3*1000*1000*1000),
+		Entry("specifying memory 3Gi", "3Gi", 3*1024*1024*1024),
+		Entry("specifying memory 45Gi", "45Gi", 45*1024*1024*1024),
+		Entry("specifying memory 2780Gi", "2780Gi", 2780*1024*1024*1024),
+		Entry("specifying memory 451231 bytes", "451231", 451231),
+	)
+	It("should calculate memory in bytes", func() {
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: Test name should be `should fail when memory size is negative`

Also please assert the domain is equal to empty domain.

```
	Expect(domain).To(Equal(api.Domain{}))
```

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+		vmi := libvmi.New(
+			libvmi.WithMemoryRequest("-45Gi"),
+		)
+		var domain api.Domain
+		configurator := compute.MemoryConfigurator{}
+		err := configurator.Configure(vmi, &domain)
+		Expect(err).To(HaveOccurred())
+	})
+
+	Context("with v1.VirtualMachineInstance", func() {
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: The context description feels a bit too generic; it might be worth making it more specific.

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+			configurator := compute.MemoryConfigurator{}
+			Expect(configurator.Configure(vmi, &domain)).To(Succeed())
+
+			xmlBytes, err := xml.MarshalIndent(domain.Spec, "", "  ")
+			Expect(err).ToNot(HaveOccurred())
+
+			xml := string(xmlBytes)
+
+			Expect(strings.Contains(xml, `<memory unit="b">2222222</memory>`)).To(BeTrue(), xml)
+		})
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: Test is overlapping with `should calculate memory in bytes`, can be dropped

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+		It("should use guest memory instead of requested memory if present", func() {
+			vmi := libvmi.New(
+				libvmi.WithGuestMemory("123Mi"),
+			)
+			v1.SetObjectDefaults_VirtualMachineInstance(vmi)
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: Could we avoid using SetObjectDefaults_VirtualMachineInstance here? It populates the VMI with many fields that aren’t relevant to these tests. Keeping the test setup minimal and focused on only the required fields helps avoid cases where tests pass due to unrelated defaults being applied.

Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: I have removed all calls to the `SetObjectDefaults_VirtualMachineInstance` in the newly added `memory_test.go`. The old tests still live in `converter_test.go` and continue to call this function.

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go (hunk 7 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: Thanks for the suggestion. done.

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go (hunk 4 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: This test code no longer exists in `memory_test.go`. It still lives in the old `converter_test.go` which I haven't changed.

**Comment by harshitgupta1337** [comment]
> * Let’s avoid removing tests from `converter_test` for now. They still provide valuable end-to-end coverage that helps ensure the different configurations continue to work together and serve as a sanity check during the refactor.
> * Unit tests should populate only the relevant fields on the VMI and assert against the full expected domain produced by the configurator.

Thanks for the feedback, @dasionov . For now, I have added new tests in `memory_test.go` but I have not removed the old tests from `converter_test.go`. There is some overlap between those tests.
Please let me know what you think about this approach. Thanks.

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go [hunk 11]
```diff
@@ -334,7 +334,7 @@ func QuantityToByte(quantity resource.Quantity) (api.Memory, error) {
 	}
 
 	if memorySize < 0 {
-		return api.Memory{Unit: "b"}, fmt.Errorf("Memory size '%s' must be greater than or equal to 0", quantity.String())
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:337]: I assume this is required in order to assert that the domain remains empty in the negative test.
If that’s the case, could you please add a short explanation to the commit message explaining why this is needed?

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+			Entry("maxGuest equal to guest memory", &guestMemory, &guestMemory, &guestMemory, nil),
+			Entry("maxGuest greater than guest memory", &guestMemory, &maxGuestMemory, &guestMemory, &maxGuestMemory),
+		)
+
+		DescribeTable("guest memory and resource requests/limits settings",
+			func(expectedMemoryBytes int64, opts ...libvmi.Option) {
+				var domain api.Domain
+				vmi := libvmi.New(opts...)
+
+				v1.SetObjectDefaults_VirtualMachineInstance(vmi)
```
Comment (dasionov) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: leftover?

Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: Removed it. Thanks

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go (hunk 11 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:337]: Added

**Comment by harshitgupta1337** [comment]
Hi @dasionov ,
Can you take another look at this PR again when you get a chance?

**Review by dasionov — APPROVED** [review]

**Comment by harshitgupta1337** [comment]
@dasionov @orelmisan Can you please take another look at this PR ?

File: pkg/virt-launcher/virtwrap/converter/compute/BUILD.bazel [hunk 13]
```diff
@@ -14,6 +14,7 @@ go_library(
         "hypervisor_features.go",
         "input_device.go",
         "launch_security.go",
+        "memory.go",
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/compute/BUILD.bazel:19]: Could you please add context to the commit message regarding the reasoning behind these changes? Same for the other commits.

File: pkg/virt-launcher/virtwrap/converter/converter_test.go [hunk 14]
```diff
@@ -3464,26 +3464,27 @@ var _ = Describe("Converter", func() {
 				c = &ConverterContext{
 					VirtualMachine: vmi,
 					AllowEmulation: true,
+					Architecture:   archconverter.NewConverter(runtime.GOARCH),
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/converter_test.go:3555]: This makes the tests dependent on the host architecture. Could you please elaborate on why this dependency is necessary?

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+		Expect(domain).To(Equal(expectedDomain))
+	},
+		Entry("specifying memory 64M", "64M", 64*1000*1000),
+		Entry("specifying memory 64Mi", "64Mi", 64*1024*1024),
+		Entry("specifying memory 3G", "3G", 3*1000*1000*1000),
+		Entry("specifying memory 3Gi", "3Gi", 3*1024*1024*1024),
+		Entry("specifying memory 45Gi", "45Gi", 45*1024*1024*1024),
+		Entry("specifying memory 2780Gi", "2780Gi", 2780*1024*1024*1024),
+		Entry("specifying memory 451231 bytes", "451231", 451231),
+		Entry("specifying float memory", "2222222200m", 2222222),
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go:61]: Could you please elaborate on the reasoning behind the large number of test cases for this specific logic?

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+		// Since conversion failed, domain should remain empty
+		Expect(domain).To(Equal(api.Domain{}))
+	})
+
+	Context("configure multiple memory fields", func() {
+		var guestMemory resource.Quantity = resource.MustParse("32Mi")
+		var maxGuestMemory resource.Quantity = resource.MustParse("128Mi")
+
+		DescribeTable("maxGuest and guest memory settings",
+			func(guestMemory *resource.Quantity, maxGuestMemory *resource.Quantity, expectedGuestMemory *resource.Quantity, expectedMaxMemory *resource.Quantity) {
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: Can we take in the VMI object instead?

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+				}
+
+				Expect(domain).To(Equal(expectedDomain))
+			},
+			Entry("maxGuest is missing", &guestMemory, nil, &guestMemory, nil),
+			Entry("maxGuest equal to guest memory", &guestMemory, &guestMemory, &guestMemory, nil),
+			Entry("maxGuest greater than guest memory", &guestMemory, &maxGuestMemory, &guestMemory, &maxGuestMemory),
+		)
+
+		DescribeTable("guest memory and resource requests/limits settings",
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/compute/memory_test.go:112]: Is it possible to combine this test with the previous one?

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go [hunk 18]
```diff
@@ -334,7 +334,7 @@ func QuantityToByte(quantity resource.Quantity) (api.Memory, error) {
...
+		return api.Memory{}, fmt.Errorf("Memory size '%s' must be greater than or equal to 0", quantity.String())
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:337]: Could you please elaborate on this change?

File: pkg/virt-launcher/virtwrap/converter/converter_test.go (hunk 14 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go:3555]: The `Architecture` field of the `ConverterContext` cannot remain `nil` because it is referenced by the Converter logic here:  https://github.com/kubevirt/kubevirt/blob/5322c71a87ee41bd01a611d269645d3035da9f02/pkg/virt-launcher/virtwrap/converter/converter.go#L1125

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go (hunk 15 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/compute/memory_test.go:61]: I have merely taken the memory-related tests from the `converter_test.go` file and added them in `memory_test.go` to ensure same converage when we eventually remove the tests from `converter_test.go`.
https://github.com/kubevirt/kubevirt/blob/5322c71a87ee41bd01a611d269645d3035da9f02/pkg/virt-launcher/virtwrap/converter/converter_test.go#L1536C1-L1542C62

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go (hunk 17 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/compute/memory_test.go:112]: I would suggest that we keep the two `DescribeTable`s separate because one of them is testing the logic concerning `maxGuest` and the other is testing logic concerning VMI memory resource requests/limits.
They are still related as they are part of the same `Context` but I would prefer to keep them separate.

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go (hunk 18 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:337]: When specified memory size is negative, in our tests we expect the returned Domain object to be empty. Setting `api.Memory.Unit` field leads to the Domain object being non-empty and failing the test.

File: pkg/virt-launcher/virtwrap/converter/compute/memory_test.go (hunk 16 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/compute/memory_test.go]: Done.

File: pkg/virt-launcher/virtwrap/converter/compute/BUILD.bazel (hunk 13 above)
Comment (harshitgupta1337) [reply pkg/virt-launcher/virtwrap/converter/compute/BUILD.bazel:19]: Done.
